File: main/insertion-sort-list/insertion-sort-list.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode InsertionSortList(ListNode head) {
        var destSentinel = new ListNode(0);

        while (head != null) {
            var dest = destSentinel;
            while (dest.next != null && dest.next.val < head.val)
                dest = dest.next;

            var next = head.next;
            head.next = dest.next;
            dest.next = head;

            head = next;
        }

        return destSentinel.next;
    }
}

File: main/insertion-sort-list/insertion-sort-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *insertionSortList(struct ListNode *head) {
    struct ListNode sentinel = {0, NULL};
    
    while (head) {
        // find the insertion point
        struct ListNode* dest = &sentinel;
        while (dest->next && dest->next->val < head->val)
            dest = dest->next;
        
        // insert the node and advance the unsorted head pointer
        struct ListNode* const next = head->next;
        head->next = dest->next;
        dest->next = head;
        head = next;
    }
    
    return sentinel.next;
}
